src/memory/AmdFlash.cc

```diff
@@ -652,18 +696,82 @@ bool AmdFlash::checkCommandEraseChip()
 
 			for (const Sector& sector : sectors) {
 				if (isWritable(sector)) {
-					ram->memset(sector.writeAddress, 0xff, sector.size);
+					erase.markPending(sector);
 				} else {
 					statusRegister.sectorLock = true;
 				}
 			}
 
-			status.dataPolling = true; // immediate completion
+			statusRegister.ready = false;
+			status.dataPolling = false;
+			setState(State::ERASE_CHIP);
+			scheduleEraseOperation(time);
 		}
 	}
 	return false;
 }
 
+void AmdFlash::scheduleEraseOperation(EmuTime::param time)
+{
+	if (state == State::ERASE_SECTOR && !status.eraseTimer) {
+		syncOperation.removeSyncPoint();
+		syncOperation.setSyncPoint(time + chip.erase.timerDuration);
+	} else if (std::ranges::any_of(erase.buffer, std::identity{})) {
+		assert(!syncOperation.pendingSyncPoint());
+		syncOperation.setSyncPoint(time + chip.erase.duration);
+	} else {
+		assert(!syncOperation.pendingSyncPoint());
+		syncOperation.setSyncPoint(time + chip.erase.minimumDuration);
+	}
+}
+
+void AmdFlash::execEraseOperation(EmuTime::param time)
+{
+	assert(state == one_of(State::ERASE_SECTOR, State::ERASE_CHIP));
+	if (state == State::ERASE_SECTOR && !status.eraseTimer) {
+		status.eraseTimer = true;
+		scheduleEraseOperation(time);
+	} else {
+		while (erase.index < sectors.size()) {
+			const Sector& sector = sectors[erase.index++];
+			if (erase.isPending(sector)) {
+				assert(!sector.writeProtect);
+				ram->memset(sector.writeAddress, 0xff, sector.size);
+				break;
+			}
+		}
+
+		if (erase.index >= sectors.size()) {
+			erase.reset();
+			statusRegister.ready = true;
+			status.dataPolling = true;
+			status.eraseTimer = false;
+			setState(State::IDLE);
+		} else {
+			syncOperation.setSyncPoint(time + chip.erase.duration);
+		}
+	}
+}
+
+bool AmdFlash::EraseOperation::isPending(const Sector& sector) const
+{
+	return sector.index >= index && sector.index < buffer.size() && buffer[sector.index];
+}
+
+void AmdFlash::EraseOperation::markPending(const Sector& sector)
+{
+	for (size_t i = buffer.size(); i <= sector.index; ++i) {
+		buffer.push_back(false);
+	}
```

Comment:
Can this loop be replaced with something like `buffer.resize(sector.index + 1, false)`?
Possibly with some check to only enlarge the buffer (in case `sector.index < buffer.size(), or maybe this function has a precondition that this won't happen?).

Reply:
Will do, didn’t realise `resize` had a default value parameter, I thought all values would default-initialise.